src/helper/worker_pool.hpp

```diff
@@ -3,39 +3,136 @@
 //
 // This product includes software developed at Datadog
 // (https://www.datadoghq.com/). Copyright 2021 Datadog, Inc.
-#ifndef WORKER_POOL_HPP
-#define WORKER_POOL_HPP
+
+#pragma once
 
 #include <atomic>
+#include <chrono>
 #include <condition_variable>
+#include <functional>
 #include <mutex>
+#include <optional>
+#include <queue>
 #include <thread>
 #include <utility>
-#include <vector>
-
-#include "scope.hpp"
 
 namespace dds::worker {
 
-class monitor {
+class consumer_queue;
```

Comment:
Maybe some other name. Having `consumer_queue` and `producer_queue` makes it look like there are two queues. Maybe `queue_producer` and `queue_consumer`.